src/main/java/org/truffleruby/core/hash/HashLiteralNode.java

```diff
@@ -33,7 +34,10 @@ public static HashLiteralNode create(RubyNode[] keyValues) {
         } else if (keyValues.length <= PackedHashStoreLibrary.MAX_ENTRIES * 2) {
             return new PackedHashStoreLibrary.SmallHashLiteralNode(keyValues);
         } else {
-            return new BucketsHashStore.GenericHashLiteralNode(keyValues);
+            String type = System.getProperty("BigHashStrategy");
```

Comment:
Thanks! Reading system properties deep inside non-config code like that rubbed me the wrong way, I will investiagte what the options systems you're talking about.

And thanks for the very detailed tutorial about truffle boundaries, really appreciate it!

Reply:
If you don't mind, I have some general questions about condition profiles, in decreasing order of importance :

1- If I have a non-message non-specialization plain old method, and I want to profile its conditionals, should I : 
                 1- Make a new node type especially for it ? 

                 2- Pass a condition profile down to it from its message-or- 
                     specialization callers ?

                  3- Do some other thing I don't yet know ?

(1) Seems too verbose, (2) won't scale well if you're calling 2+ levels deep, (3) might not exist.

2- Why do InlineConditionProfile require an extra argument ? this article (https://docs.oracle.com/en/graalvm/jdk/20/docs/graalvm-as-a-platform/language-implementation-framework/DSLNodeObjectInlining/) about inlined nodes seems to suggest that the methods of inlined objects become sort of like static methods, and the object state is instead kept in the parent node. Is that understanding correct-ish ?

(On a seperate note, I haven't noticed us using inlined nodes in the parts of the code I have seen, am I not paying attention or is there something preventing us from using them (e.g truffle version) ?)

3- Are inlinedConditionProfile objects maximally efficient ? do they fit 16 object-state per int ? Sounds like something Truffle engineers would already think of but it never hurts to check and report if not.

Reply:
1. If it's just a single ConditionProfile it's often OK to just pass it as an argument, if it's more a inlineable node is better as it encapsulates those details.
2.
> Is that understanding correct-ish ?

Yes-ish. The methods don't move (they can't) but the state/fields are moved to the parent node.

> (On a seperate note, I haven't noticed us using inlined nodes in the parts of the code I have seen, am I not paying attention or is there something preventing us from using them (e.g truffle version) ?)

We use it in quite a few places but not everywhere, it's a lot of work to migrate but for new nodes we should always consider it.

3. Yes, they only use 2 bits, see InlinedConditionProfile#REQUIRED_STATE_BITS

Reply:
Regarding the system property, you could just read it in a static initializer and then store the result in a `static final` field.
Because it's meant to be a temporary option (we'd just remove the old Buckets impl at the end), and on native language options would cause to skip the pre-initialized context so a language option wouldn't be a good fit anyway (and a context option would incur overhead).